File: strategy-pattern/java/src/ContextDog.java
package src;

public class ContextDog extends Context {

  public ContextDog() {
    // 使用某个策略
    System.out.println("ContextDog::setStrategy(StrategyB).");
    this.setStrategy(new StrategyB());
  }

  // 这样就是强制绑定了策略动作，不够灵活
  // @Override
  // public void action() {
  //   new StrategyC().run();
  // }

}

File: strategy-pattern/go/src/context_dog.go
package src

import "fmt"

// 定义具体执行对象，Go没有继承，用聚合来调用Context里的函数
type ContextDog struct {
	context Context
}

// 可提前绑定具体的策略
func (c *ContextDog) Init() {
	c.context.SetStrategy(&StrategyB{})
	fmt.Println("ContextDog::init. setStrategy(StrategyB)")
}

// 调用策略方法
func (c *ContextDog) Run() {
	fmt.Println("ContextDog::run")
	c.context.Run()
}
